File: packages/app/src/services/software/definitions/vivaldi.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'vivaldi';
	public name = 'Vivaldi';
	public category = [SoftwareCategory.Browser];
	public downloadName = 'Vivaldi.x64.exe';
	public shouldCacheUrl = true;
	public icon = 'vivaldi.png';
	public homepage = 'https://vivaldi.com';

	public async resolveDownloadUrl() {
		const res = await fetch('https://vivaldi.com/download');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/class="download-button download-link" href="(https:\/\/downloads\.vivaldi\.com\/stable\/Vivaldi\.\d+\.\d+\.\d+\.\d+\.x64\.exe)"/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[1]);
	}
}

File: src/lib/software/vivaldi.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class Vivaldi(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'vivaldi'
        self.name = 'Vivaldi'
        self.category = [SoftwareCategory.Browser]
        self.download_name = 'Vivaldi.x64.exe'
        self.icon = 'vivaldi.png'
        self.homepage = 'https://vivaldi.com'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'class="download-button download-link" href="(https://downloads\.vivaldi\.com/stable/Vivaldi\.\d+\.\d+\.\d+\.\d+\.x64\.exe)"')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(1))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://vivaldi.com/download/'))
